packages/react-meteor-data/useTracker.js

```diff
@@ -0,0 +1,155 @@
+    return false;
+  }
+
+  if (!Array.isArray(nextDeps)) {
+    if (Meteor.isDevelopment) {
+      warn(
+        'Warning: useTracker expected an dependency value of '
+        + `type array but got type of ${typeof nextDeps} instead.`
+      );
+    }
+    return false;
+  }
+
+  const len = nextDeps.length;
+
+  if (prevDeps.length !== len) {
+    return false;
+  }
+
+  for (let i = 0; i < len; i++) {
+    if (!is(nextDeps[i], prevDeps[i])) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Used to create a forceUpdate from useReducer. Forces update by
+// incrementing a number whenever the dispatch method is invoked.
+const fur = x => x + 1;
+
+function useTracker(reactiveFn, deps) {
+  const { current: refs } = useRef({});
+
+  const [, forceUpdate] = useReducer(fur, 0);
+
+  const dispose = () => {
+    if (refs.computation) {
+      refs.computation.stop();
+      refs.computation = null;
+    }
+  };
+
+  // this is called like at componentWillMount and componentWillUpdate equally
+  // in order to support render calls with synchronous data from the reactive computation
+  // if prevDeps or deps are not set areHookInputsEqual always returns false
+  // and the reactive functions is always called
+  if (!areHookInputsEqual(deps, refs.previousDeps)) {
+    // if we are re-creating the computation, we need to stop the old one.
+    dispose();
+
+    // store the deps for comparison on next render
+    refs.previousDeps = deps;
+
+    // Use Tracker.nonreactive in case we are inside a Tracker Computation.
+    // This can happen if someone calls `ReactDOM.render` inside a Computation.
+    // In that case, we want to opt out of the normal behavior of nested
+    // Computations, where if the outer one is invalidated or stopped,
+    // it stops the inner one.
+    refs.computation = Tracker.nonreactive(() => (
+      Tracker.autorun((c) => {
+        const runReactiveFn = () => {
+          const data = reactiveFn();
+          if (Meteor.isDevelopment) checkCursor(data);
+          refs.trackerData = data;
+        };
+
+        if (c.firstRun) {
+          // This will capture data synchronously on first run (and after deps change).
+          // Additional cycles will follow the normal computation behavior.
+          runReactiveFn();
+        } else {
+          // If deps are falsy, stop computation and let next render handle reactiveFn.
+          if (refs.previousDeps !== null && refs.previousDeps !== undefined
+            && !Array.isArray(refs.previousDeps)) {
+            dispose();
+          } else {
+            runReactiveFn();
+          }
+          forceUpdate();
+        }
+      })
+    ));
+  }
+
+  // stop the computation on unmount only
+  useEffect(() => {
+    if (Meteor.isDevelopment
+      && deps !== null && deps !== undefined
+      && !Array.isArray(deps)) {
+      warn(
+        'Warning: useTracker expected an initial dependency value of '
```

Comment:
You are right, this is confusing. I'm not even sure it's correct. Verifying this is one of the reasons I asked about the computation life-cycle argument.

Reply:
I clarified some of this in recent commits/comments. I think it's even cleaner in hooks-cleanup-2 branch

Reply:
Would still be more legible if this was moved out of the way and into a Meteor.isDevelopment && checkDeps(deps) ?

Reply:
Really all that's needed for most of these (some of them are slightly different) is `!Array.isArray(deps)` - checking for null and undefined is a performance optimization, which wouldn't be optimized if we used a function. For readability, just using `Array.isArray` is best - but I figured the optimization was worth it?

Reply:
sure, why not the optimization - but in itself it would deserve a comment because yes, anyone looking at this code is very likely to wonder "why not just Array.isArray() ?", and this is another argument for splitting that to a function, because we don't want to duplicate the comment as well :-D

Reply:
I actually did change this particular check:

```js
  // stop the computation on unmount only
  useEffect(() => {
    // falsy deps is okay, but if deps is not falsy, it must be an array
    if (Meteor.isDevelopment && (deps && !Array.isArray(deps))) {
      warn(
        'Warning: useTracker expected an initial dependency value of '
        + `type array, null or undefined but got type of ${typeof deps} instead.`
      );
    }

```

Reply:
Check the new PR #273, this is resolved differently